_rules/table-header-cells-refer-to-data-cells-d0f69e.md

```diff
@@ -0,0 +1,201 @@
+---
+id: d0f69e
+name: All table header cells have corresponding data cells
+rule_type: atomic
+description: |
+  This rule checks that each table header has corresponding data cells in a table element.
+accessibility_requirements:
+  wcag20:1.3.1: # Info and Relationships (A)
+    forConformance: true
+    failed: not satisfied
+    passed: further testing needed
+    inapplicable: further testing needed
+input_aspects:
+  - DOM Tree
+authors:
+  - Jey Nandakumar
+  - Audrey Maniez
+---
+
+## Applicability
+
+The rule applies to any HTML element that is a [descendant](https://dom.spec.whatwg.org/#concept-tree-descendant) in the [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree) of an element having a [semantic role](#semantic-role) of either [table](https://www.w3.org/TR/wai-aria-1.1/#table) or [grid](https://www.w3.org/TR/wai-aria-1.1/#grid), and is [visible](#visible) as well as [included in the accessibility tree](#included-in-the-accessibility-tree), where the element has any of the following [semantic roles](#semantic-role):
+
+- [cell](https://www.w3.org/TR/wai-aria-1.1/#cell)
+- [gridcell](https://www.w3.org/TR/wai-aria-1.1/#gridcell)
+- [rowheader](https://www.w3.org/TR/html-aria/#index-aria-rowheader)
+- [columnheader](https://www.w3.org/TR/html-aria/#index-aria-columnheader)
+
+## Expectation
+
+The target element is either the row or column header for cells that is a [descendant](https://dom.spec.whatwg.org/#concept-tree-descendant) in the [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree) of the same [table](https://www.w3.org/TR/wai-aria-1.1/#table) or [grid](https://www.w3.org/TR/wai-aria-1.1/#grid) element, as that of the target element.
+
+**Note:** Assigning headers cells to data cell is performed as per algorithm - [Forming relationships between data cells and header cells](https://html.spec.whatwg.org/multipage/tables.html#header-and-data-cell-semantics).
+
+## Assumptions
+
+_There are currently no assumptions._
+
+## Accessibility Support
+
+_There are no major accessibility support issues known for this rule._
+
+## Background
+
+- [Understanding Success Criterion 1.3.1: Information and relationships](https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships.html)
+- [H43: Using id and headers attributes to associate data cells with header cells in data tables](https://www.w3.org/WAI/WCAG21/Techniques/html/H43)
+- [Forming relationships between data cells and header cells](https://html.spec.whatwg.org/multipage/tables.html#header-and-data-cell-semantics)
+
+## Test Cases
```

Comment:
Examples to add:
* `headers` attribute giving a cell to an otherwise empty header (should pass)
```html
<table>
  <tr> <th id="foo"></th> <th id="bar"></th> </tr>
  <tr> <td></td> </tr>
  <tr> <td headers="bar"></td> </tr>
</table>
```

```html
<table>
  <tr> <th id="foo"></th> <th id="bar"></th> </tr>
  <tr> <td headers="foo bar"></td> </tr>
</table>
```

* `headers` attribute removing cell from a header (should fail)
```html
<table>
  <tr> <th id="foo"></th> <th id="bar"></th> </tr>
  <tr> <td></td> <td headers="foo"></td> </tr>
</table>
```

```html
<span id="foo">
<table>
  <tr> <th></th> </tr>
  <tr> <td headers="foo"></td> </tr>
</table>
```

Reply:
The [algorithm for assigning headers cells](https://html.spec.whatwg.org/multipage/tables.html#header-and-data-cell-semantics) is somewhat complex, but Step 3 is very clear about one thing: if there is a `headers` attribute, its content, and only it (even if it result in am empty list of headers), is taken into consideration; if there is no `headers` attribute, the cell position, and only it, it take into consideration.

As per this algorithm, the last two examples above have header cells with no data cell assigned to it.

So, if this rule consider that "is the header for cells" is the same thing as "as some cells assigned by this algorithm", then the last two examples above fail. Otherwise, we need to define what we mean by "is the header for cells".